Create simple documentation for @typeofweb/schema library (#1)
typeofweb/schema: README.md

    - [nil](#nil)

## Introduction

@typeofweb/schema is lightweight, extensible library for data validation.

## Installation

Install package with npm:

`npm install @typeofweb/schema`

Install package with yarn:

`yarn add @typeofweb/schema`

## Example

```js
const personSchema = object({
  name: string(),
  age: number(),
  email: optional(string()),
});

const mark = {
  name: 'Mark',
  age: 29,
};

const personValidator = validate(personSchema);

// If validation is successful returns data with correct type, throws error otherwise
const validatedPerson = personValidator(mark);
```

## API

### oneOf

Generates a schema which will match only specified values.

```js
const fishSchema = oneOf(['trout', 'catfish']);
const fishValidator = validate(fishSchema);

// Returns 'trout'
const trout = fishValidator('trout');

// Throws validation error
const salmon = fishValidator('salmon');
```

### string

Generates schema that will match string.

```js
const stringSchema = string();
const stringValidator = validate(stringSchema);

// Return 'Micheal'
const micheal = stringValidator('Micheal');
```

### number

Generates schema that will match number.

```js

> const numberSchema = string();

Reviewer: ```suggestion
const numberSchema = number();
```

const numberValidator = validate(numberSchema);

// Return 3.14
const pi = numberValidator(3.14);
```

### boolean

Generates schema that will match boolean.

```js

> const booleanSchema = string();

Reviewer: ```suggestion
const booleanSchema = boolean();
```

const booleanValidator = validate(booleanSchema);

// Returns true
const loading = booleanValidator(true);
```

### date

Generates schema that will match date.

```js
const dateSchema = date();
const dateValidator = validate(dateSchema);

// Returns Date Tue Jan 12 2021 22:00:20 GMT+0100 (Central European Standard Time)
const annieBirthday = dateValidator(
  new Date('Tue Jan 12 2021 22:00:20 GMT+0100 (Central European Standard Time)'),
);
```

### object

Generates schema that will match object.

```js
const carSchema = object({
  manufacturer: string(),
  model: string(),
  mass: number(),
  enginePower: number(),
  fuelCapacity: number(),
});
const carValidator = validate(carSchema);

/* Returns {
  manufacturer: 'Typeofweb enterprise',
  model: 'x1024',
  mass: 1600,
  enginePower: 135,
  fuelCapacity: 59,
} */
const matthewCar = carValidator({
  manufacturer: 'Typeofweb enterprise',
  model: 'x1024',
  mass: 1600,
  enginePower: 135,
  fuelCapacity: 59,
});
```

### array

> Generates schema that will match all values specifiec in the schemas in the array.

Reviewer: ```suggestion
Generates schema that will match all values specified in the schemas in the array.
```

// Returns ['classical', 'lofi', 'pop']
const musicGenres = musicGenresValidator(['classical', 'lofi', 'pop']);
```

### Modifiers

Modifers are used to customize the schema and change the validation behaviour.

#### nullable

If you use nullable modifier, output type will be `null | InferredType`.

```js
const nullableRoleSchema = nullable(oneOf(['User', 'Admin']));
const nullableRoleValidator = validate(nullableRoleSchema);

// Returns 'User', the output type is null | 'User' | 'Admin'
const role = nullableRoleValidator('User');
```

### optional

If you use optional modifier, output type will be `undefined | InferredType`.

```js
const optionalRoleSchema = optional(oneOf(['User', 'Admin']));
const optionalRoleValidator = validate(optionalRoleSchema);

// Returns 'User', the output type is undefined | 'User' | 'Admin'
const role = optionalRoleValidator('User');
```

### nil

If you use nil modifier, output type will be `undefined | null | InferredType`.

```js
const nilRoleSchema = nil(oneOf(['User', 'Admin']));
const nilRoleValidator = validate(nilRoleSchema);

// Returns 'User', the output type is undefined | null | 'User' | 'Admin'
const role = nilRoleValidator('User');

> ```

Reviewer: It would be nice to mention that null/undefined also pass the validation

[![codecov](https://codecov.io/gh/typeofweb/schema/branch/main/graph/badge.svg?token=6DNCIHEEUO)](https://codecov.io/gh/typeofweb/schema)
[![npm](https://img.shields.io/npm/v/@typeofweb/schema.svg)](https://www.npmjs.com/package/@typeofweb/schema)
[![npm bundle size (minified + gzip)](https://img.shields.io/bundlephobia/minzip/@typeofweb/schema.svg)](https://www.npmjs.com/package/@typeofweb/schema)

## Table of contents

- [Introduction](#introduction)
- [Installation](#installation)
- [Example](#example)
- [API](#api)
  - [oneOf](#oneOf)
  - [string](#string)
  - [number](#number)
  - [boolean](#boolean)
  - [date](#date)
  - [object](#object)
  - [array](#array)
  - [Modifiers](#modifiers)
    - [nullable](#nullable)
    - [optional](#optional)
    - [nil](#nil)

## Introduction

> @typeofweb/schema is lightweight, extensible library for data validation.

Reviewer: ```suggestion
@typeofweb/schema is a lightweight and extensible library for data validation with full support for TypeScript!
```

## Installation

> Install package with npm:

Reviewer: ```suggestion
Install the package with npm:
```

`npm install @typeofweb/schema`

> Install package with yarn:

Reviewer: ```suggestion
or install the package with yarn:
```

`yarn add @typeofweb/schema`

## Example

```js
const personSchema = object({
  name: string(),
  age: number(),
  email: optional(string()),
});

const mark = {
  name: 'Mark',
  age: 29,
};

const personValidator = validate(personSchema);

// If validation is successful returns data with correct type, throws error otherwise
const validatedPerson = personValidator(mark);
```

## API

### oneOf

> Generates a schema which will match only specified values.

Reviewer 2: ```suggestion
Generates a schema which will match only the specified values.
```

Reviewer: To chyba nie jest poprawnie? :P

> Generates a schema which will match only specified values.

Reviewer: ```suggestion
Generates a schema which matches only specified values.
```

> ```js

Reviewer: ```suggestion
```ts
```

const fishSchema = oneOf(['trout', 'catfish']);
const fishValidator = validate(fishSchema);

// Returns 'trout'
const trout = fishValidator('trout');

> // Throws validation error

Reviewer: ```suggestion
// Throws ValidationError
```

const salmon = fishValidator('salmon');
```

### string

> Generates schema that will match string.

Reviewer 2: ```suggestion
Generates schema that will match a string.
```

> Generates schema that will match string.

Reviewer: ```suggestion
Creates a schema that matches strings.
```

> ```js

Reviewer: ```suggestion
```ts
```

const stringSchema = string();
const stringValidator = validate(stringSchema);

// Return 'Micheal'
const micheal = stringValidator('Micheal');
```

### number

> Generates schema that will match number.

Reviewer 2: ```suggestion
Generates schema that will match a number.
```

> Generates schema that will match number.

Reviewer: ```suggestion
Creates a schema that matches numbers.
```

> ```js

Reviewer: ```suggestion
```ts
```

const numberSchema = number();
const numberValidator = validate(numberSchema);

> // Return 3.14

Reviewer: ```suggestion
// Returns 3.14
```

const pi = numberValidator(3.14);
```

### boolean

> Generates schema that will match boolean.

Reviewer 2: ```suggestion
Generates schema that will match a boolean.
```

> Generates schema that will match boolean.

Reviewer: ```suggestion
Creates a schema that matches booleans.
```

> ```js

Reviewer: ```suggestion
```ts
```

const booleanSchema = boolean();
const booleanValidator = validate(booleanSchema);

// Returns true
const loading = booleanValidator(true);
```

### date

> Generates schema that will match date.

Reviewer 2: ```suggestion
Generates schema that will match a JavaScript Date object.
```

> Generates schema that will match date.

Reviewer: ```suggestion
Creates a schema that matches JavaScript `Date` objects.
```

> ```js

Reviewer: ```suggestion
```ts
```

const dateSchema = date();
const dateValidator = validate(dateSchema);

// Returns Date Tue Jan 12 2021 22:00:20 GMT+0100 (Central European Standard Time)
const annieBirthday = dateValidator(
  new Date('Tue Jan 12 2021 22:00:20 GMT+0100 (Central European Standard Time)'),
);
```

### object

> Generates schema that will match object.

Reviewer 2: ```suggestion
Generates schema that will match a object.
```

> Generates schema that will match object.

Reviewer: ```suggestion
Creates a schema that matches objects of provided shape.
```

> ```js

Reviewer: ```suggestion
```ts
```

const carSchema = object({
  manufacturer: string(),
  model: string(),
  mass: number(),
  enginePower: number(),
  fuelCapacity: number(),
});
const carValidator = validate(carSchema);

/* Returns {
  manufacturer: 'Typeofweb enterprise',
  model: 'x1024',
  mass: 1600,
  enginePower: 135,
  fuelCapacity: 59,
} */
const matthewCar = carValidator({

>   manufacturer: 'Typeofweb enterprise',

Reviewer: ```suggestion
  manufacturer: 'Type of Web Enterprise',
```

  model: 'x1024',
  mass: 1600,
  enginePower: 135,
  fuelCapacity: 59,
});
```

### array

> Generates schema that will match all values specified in the schemas in the array.

Reviewer 2: ```suggestion
Generates a schema that will match all the values specified by the schemas passed in the array.
```

> Generates schema that will match all values specified in the schemas in the array.

Reviewer: ```suggestion
Creates a schema that matches arrays containing values specified by the schemas passed in the array.
```

> ```js

Reviewer: ```suggestion
```ts
```

const musicGenresSchema = array([string()]);
const musicGenresValidator = validate(musicGenresSchema);

// Returns ['classical', 'lofi', 'pop']
const musicGenres = musicGenresValidator(['classical', 'lofi', 'pop']);

> ```

Reviewer: We could use an example with more than one schema inside the array

### Modifiers

> Modifers are used to customize the schema and change the validation behaviour.

Reviewer: ```suggestion
Modifiers are used to customize schemas and change their validation behaviour.
```

#### nullable

> If you use nullable modifier, output type will be `null | InferredType`.

Reviewer: ```suggestion
If you use the `nullable` modifier, given schema will be extended to also match `null` values, and its output type will be `null | T`.
```

> ```js

Reviewer: ```suggestion
```ts
```

const nullableRoleSchema = nullable(oneOf(['User', 'Admin']));
const nullableRoleValidator = validate(nullableRoleSchema);

// Returns 'User', the output type is null | 'User' | 'Admin'
const role = nullableRoleValidator('User');

// It's also fine
nullableRoleValidator(null);

> // Throws validation error

Reviewer: ```suggestion
// Throws ValidationError
```

nullableRoleValidator();
nullableRoleValidator(undefined);
```

### optional

> If you use optional modifier, output type will be `undefined | InferredType`.

Reviewer: ```suggestion
If you use the `optional` modifier, given schema will be extended to also match `undefined` values, and its output type will be `undefined | T`.
```

> ```js

Reviewer: ```suggestion
```ts
```

const optionalRoleSchema = optional(oneOf(['User', 'Admin']));
const optionalRoleValidator = validate(optionalRoleSchema);

// Returns 'User', the output type is undefined | 'User' | 'Admin'
const role = optionalRoleValidator('User');

// It's also fine
optionalRoleValidator();
optionalRoleValidator(undefined);

// Throws validation error
optionalRoleValidator(null);
```

### nil

> If you use nil modifier, output type will be `undefined | null | InferredType`.

Reviewer: ```suggestion
If you use the `nil` modifier, given schema will be extended to also match `null` and `undefined` values, and its output type will be `null | undefined | T`.
```

> ```js

Reviewer: ```suggestion
```ts
```
